samples/bluetooth/handsfree_ag/src/main.c

```diff
@@ -0,0 +1,386 @@
+}
+
+static void ag_incoming(struct bt_hfp_ag *ag, const char *number)
+{
+	printk("Incoming call, remote number %s\n", number);
+	k_work_cancel_delayable(&call_connect_work);
+}
+
+static struct bt_hfp_ag_cb ag_cb = {
+	.connected = ag_connected,
+	.disconnected = ag_disconnected,
+	.sco_connected = ag_sco_connected,
+	.sco_disconnected = ag_sco_disconnected,
+	.outgoing = ag_outgoing,
+	.incoming = ag_incoming,
+	.ringing = ag_ringing,
+	.accept = ag_accept,
+	.reject = ag_reject,
+	.terminate = ag_terminate,
+};
+
+static uint8_t sdp_discover_cb(struct bt_conn *conn, struct bt_sdp_client_result *result)
+{
+	int err;
+	uint16_t value;
+
+	printk("Discover done\n");
+
+	if (result->resp_buf != NULL) {
+		err = bt_sdp_get_proto_param(result->resp_buf, BT_SDP_PROTO_RFCOMM, &value);
+
+		if (err != 0) {
+			printk("Fail to parser RFCOMM the SDP response!\n");
+		} else {
+			printk("The server channel is %d\n", value);
+			err = bt_hfp_ag_connect(conn, &hfp_ag, value);
+			if (err != 0) {
+				printk("Fail to create hfp AG connection (err %d)\n", err);
+			}
+		}
+	}
+
+	return BT_SDP_DISCOVER_UUID_STOP;
+}
+
+static struct bt_sdp_discover_params sdp_discover = {
+	.func = sdp_discover_cb,
+	.pool = &sdp_discover_pool,
+	.uuid = BT_UUID_DECLARE_16(BT_SDP_HANDSFREE_SVCLASS),
+};
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	int res;
+
+	if (err) {
+		if (default_conn != NULL) {
+			default_conn = NULL;
+		}
+		printk("Connection failed (err 0x%02x)\n", err);
+	} else {
+		if (default_conn == conn) {
+			struct bt_conn_info info;
+
+			bt_conn_get_info(conn, &info);
+			if (info.type != BT_CONN_TYPE_BR) {
+				return;
+			}
+
+			/*
+			 * Do an SDP Query on Successful ACL connection complete with the
+			 * required device
+			 */
+			res = bt_sdp_discover(default_conn, &sdp_discover);
+			if (res) {
+				printk("SDP discovery failed (err %d)\r\n", res);
+			} else {
+				printk("SDP discovery started\r\n");
+			}
+			printk("Connected\n");
+		}
+	}
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	printk("Disconnected (reason 0x%02x)\n", reason);
+
+	if (default_conn != conn) {
+		return;
+	}
+
+	if (default_conn) {
+		default_conn = NULL;
+	} else {
+		return;
+	}
+}
+
+static void security_changed(struct bt_conn *conn, bt_security_t level, enum bt_security_err err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	struct bt_conn_info info;
+
+	bt_conn_get_info(conn, &info);
+
+	bt_addr_to_str(info.br.dst, addr, sizeof(addr));
+
+	printk("Security changed: %s level %u (err %d)\n", addr, level, err);
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed,
+};
+
+static void scan_discovery_cb(struct bt_br_discovery_result *results, size_t count)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	uint8_t *eir;
+	bool flag = true;
+	static uint8_t temp[240];
+	size_t len = sizeof(results->eir);
+	size_t major_device;
+	size_t minor_device;
```

Comment:
`size_t` doesn't sound like the right type for this. It's supposed to be used (as the name suggests) to indicate the size of something. Maybe `uint8_t` would be more suitable.

Reply:
updated